File: Python/ExampleTests/common/patch_title_typed.py
from . import json_util as _json
from .patch_title_string import PatchTitleString

class PatchTitleTyped:
    def __init__(self, title: str = ''):
        self.title = title

    @classmethod
    def from_string_obj(cls, s: PatchTitleString) -> 'PatchTitleTyped':
        return cls(
            s.title
        )

    def to_string_obj(self) -> PatchTitleString:
        return PatchTitleString(
            str(self.title)
        )

    @staticmethod
    def to_string_list(items) -> list:
        return [t.to_string_obj() for t in items]

    @staticmethod
    def from_string_list(items) -> list:
        return [PatchTitleTyped.from_string_obj(s) for s in items]

    def to_json_value(self) -> dict:
        return {
            'title': self.title,
        }

    def to_json(self) -> str:
        return _json.dumps(self.to_json_value())

    @classmethod
    def from_json_value(cls, m: dict) -> 'PatchTitleTyped':
        return cls(
            _json.as_str(_json.require(m, 'title'), 'title')
        )

    @classmethod
    def from_json(cls, text: str) -> 'PatchTitleTyped':
        return cls.from_json_value(_json.loads(text))

    @staticmethod
    def to_json_list(items) -> str:
        return _json.dumps([item.to_json_value() for item in items])

    @classmethod
    def from_json_list(cls, text: str) -> list:
        raw = _json.as_list(_json.loads(text), 'PatchTitleTyped')
        return [cls.from_json_value(e) for e in raw]

    def __str__(self):
        return (f'title={self.title}')

    def _key(self):
        return (self.title,)

    def __eq__(self, other):
        if not isinstance(other, PatchTitleTyped):
            return NotImplemented
        return self._key() == other._key()

    def __hash__(self):
        return hash(self._key())

    def __repr__(self):
        return f'PatchTitleTyped({self})'

File: JavaScript/ExampleTests/common/PatchTitleTyped.js
import { PatchTitleString } from "./PatchTitleString.js";
import * as _json from "./json.js";

export class PatchTitleTyped {
  constructor(title = "") {
    this.title = title;
  }

  static fromStringObj(s) {
    return new PatchTitleTyped(
      s.title
    );
  }

  toStringObj() {
    return new PatchTitleString(
      String(this.title)
    );
  }

  static toStringList(list) {
    return list.map(t => t.toStringObj());
  }

  static fromStringList(list) {
    return list.map(s => PatchTitleTyped.fromStringObj(s));
  }

  toJsonValue() {
    return {
      title: this.title,
    };
  }

  toJSON() { return _json.stringify(this.toJsonValue()); }

  static fromJsonValue(m) {
    return new PatchTitleTyped(
      _json.asString(_json.require(m, "title"), "title")
    );
  }

  static fromJSON(text) { return PatchTitleTyped.fromJsonValue(_json.parse(text)); }

  static toJSONList(items) {
    return _json.stringify(items.map((item) => item.toJsonValue()));
  }

  static fromJSONList(text) {
    const raw = _json.asArray(_json.parse(text), "PatchTitleTyped");
    return raw.map((e) => PatchTitleTyped.fromJsonValue(e));
  }

  toString() {
    return `title=${this.title}`;
  }

  equals(other) {
    if (!(other instanceof PatchTitleTyped)) return false;
    return this.title === other.title;
  }
}
